addons/persistence/postgresql-persistence-addon/src/main/java/org/kie/kogito/persistence/postgresql/PostgreProcessInstances.java

```
@@ -0,0 +1,282 @@
+        updateInternal(UUID.fromString(id), marshaller.marshallProcessInstance(instance));
+        disconnect(instance);
+    }
+
+    @Override
+    public void remove(String id) {
+        deleteInternal(UUID.fromString(id));
+    }
+
+    @Override
+    public Optional<ProcessInstance> findById(String id, ProcessInstanceReadMode mode) {
+        return findByIdInternal(UUID.fromString(id)).map(b -> marshaller.unmarshallProcessInstance(b, process));
+    }
+
+    @Override
+    public Collection<ProcessInstance> values(ProcessInstanceReadMode mode) {
+        return findAllInternal().stream().map(i -> marshaller.unmarshallProcessInstance(i, process)).collect(Collectors.toList());
+    }
+
+    @Override
+    public Integer size() {
+        return countInternal().intValue();
+    }
+
+    private void disconnect(ProcessInstance instance) {
+        ((AbstractProcessInstance<?>) instance).internalRemoveProcessInstance(() -> {
+            try {
+                byte[] reloaded = findByIdInternal(UUID.fromString(instance.id())).get();
+                return marshaller.unmarshallWorkflowProcessInstance(reloaded, process);
+            } catch (RuntimeException e) {
+                LOGGER.error("Unexpected exception thrown when reloading process instance {}", instance.id(), e);
+                return null;
+            }
+        });
+    }
+
+    private boolean insertInternal(UUID id, byte[] payload) {
+        try {
+            final CompletableFuture<RowSet<Row>> future = new CompletableFuture<>();
+            client.preparedQuery("INSERT INTO process_instances (id, payload, process_id) VALUES ($1, $2, $3)")
+                    .execute(Tuple.of(id, Buffer.buffer(payload), process.id()), getAsyncResultHandler(future));
+            RowSet<Row> rows = getResultFromFuture(future);
+            return rows.rowCount() == 1;
+        } catch (Exception e) {
+            LOGGER.error("Error inserting process instance {}", id, e);
```

same here, we should probably propagate the exception.